### ISSUE TITLE
We should implement multithreaded DETECT_PERIODICALLY

### ISSUE THREAD

**Issue by ayberkozgur** [chili-epfl/chilitags#63]
We agree that there should be 3 ways to use Chilitags:
1. Detect all the time, simplest.
2. Detect and track all the time, takes more time, more robust, OK.
3. Detect periodically. 

Now, the current periodic detection is "1 call detect, 14 calls track, and so on". This must run on the same thread on the user side as the user cannot determine whether Chilitags will be tracking or detecting (without counting the frames manually or calling with `TRACK_ONLY` and `DETECT_ONLY`, but neither is the point of `DETECT_PERIODICALLY`; and this wouldn't be thread safe without appropriate locks). As these run in the same thread, detection delays the pose update of tags significantly; assume that detection takes 200 ms while tracking takes 15 ms. A 200 ms call every half a second causes hiccups even if the visual camera image update is running on another thread. 

Now, assume that detection and tracking run on separate threads, once one detect cycle is sacrificed, Chilitags can run at 30 FPS since the detect cycle can catch up with the track cycle. 

The easiest way to do this is to launch an additional "detection" thread when Chilitags is created. During a normal `Chilitags3D::estimate()` call, the frame is copied for the use of tracking and detection. The detection thread, having a new frame, updates the "detected" tag poses as soon as it can. Meanwhile, the `Chilitags3D::estimate()` call, after copying the frame, takes the most recent detected tag poses and uses it to track and return the results. I'm assuming that we place mutexes where necessary.

At this point, we have alternatives to choose from for various performance presets:
- Run detection every N frames completed of tracking, if possible; and as long as there's a new camera frame
- Run detection as frequently as possible, as long as there's a new camera frame

In short, I see no "correct" way of implementing `DETECT_PERIODICALLY` without at least 1 additional thread. The above implementation can be done with `pthread`s for cross-platform compatibility.

Or, another solution might be to remove `DETECT_PERIODICALLY` altogether, lock the appropriate members (such as image buffers and tag poses) across calls to `estimate()` with mutexes and expect the user to implement periodic detection with `DETECT_ONLY` and `TRACK_ONLY`. 

What do you think?
